### PR TITLE
Better error handling when the replication slot no longer exists

### PR DESCRIPTION
**Pull request by acco**

### CODE REVIEW

**Comment by acco** [comment]
* **#233** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/233?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @acco and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: lib/sequin/databases/databases.ex [hunk 1]
```diff
@@ -198,65 +197,13 @@ defmodule Sequin.Databases do
 
   def test_slot_permissions(%PostgresDatabase{} = database, %PostgresReplicationSlot{} = slot) do
     with_uncached_connection(database, fn conn ->
-      with :ok <- check_replication_slot_exists(conn, slot.slot_name),
-           :ok <- check_publication_exists(conn, slot.publication_name) do
-        check_replication_permissions(conn)
+      with :ok <- Postgres.check_replication_slot_exists(conn, slot.slot_name),
+           :ok <- Postgres.check_publication_exists(conn, slot.publication_name) do
+        Postgres.check_replication_permissions(conn)
```
Comment (acco) [inline lib/sequin/databases/databases.ex:202]: Move to Postgres, so we can share some of them with other parts of system (in this case Replication)

File: lib/sequin/postgres.ex [hunk 2]
```diff
@@ -125,6 +126,61 @@ defmodule Sequin.Postgres do
     end)
   end
 
+  def check_replication_slot_exists(conn, slot_name) do
```
Comment (acco) [inline lib/sequin/postgres.ex:129]: just moved

File: lib/sequin/postgres.ex [hunk 3]
```diff
@@ -125,6 +126,61 @@ defmodule Sequin.Postgres do
...
+      {:ok, %{num_rows: 0}} ->
+        {:error, Error.validation(summary: "Replication slot '#{slot_name}' does not exist")}
+
+      {:error, %Postgrex.Error{} = error} ->
+        {:error, ValidationError.from_postgrex("Failed to check replication slot: ", error)}
+    end
+  rescue
+    error in [DBConnection.ConnectionError] ->
+      {:error, Error.validation(summary: "Failed to check replication slot: #{error.message}")}
+  end
```
Comment (acco) [inline lib/sequin/postgres.ex:145]: added this to make more resilient

File: lib/sequin/extensions/replication.ex [hunk 4]
```diff
@@ -125,19 +126,49 @@ defmodule Sequin.Extensions.Replication do
...
+    Task.async(fn ->
+      if env() == :test and not is_nil(state.test_pid) do
+        send(state.test_pid, {:stop_replication, state.id})
+      else
+        ReplicationRuntime.Supervisor.stop_replication(state.id)
+      end
+    end)
+
+    {:noreply, state}
+  end
```
Comment (RTLS) [inline lib/sequin/extensions/replication.ex:157]: how quickly does this handle_connect get called? does it "wait" for failures before retrying? I saw it get spammed quickly in the logs once. just want to make sure we don't kill this process prematurely

**Pull request merged**